scripts/mirna_extension.py

```diff
@@ -1,240 +1,375 @@
-        extend_mirnas:
-            extends the start and end of the mature miRNAs by n nucleotides
+        db: In-memory database of the input GFF3 records.
+        db_out: In-memory database of the updated GFF3 records.
+        seq_lengths: Dictionary mapping reference sequence IDs to their
+            lengths.
     """
 
     def __init__(self) -> None:
         """Initialize class."""
-        self.db = None
-
-    def load_gff_file(self, gff_file: Optional[Path] = None) -> None:
-        """Load GFF3 file.
+        self.db: Optional[gffutils.FeatureDB] = None
+        self.db_out: Optional[gffutils.FeatureDB] = None
+        self.seq_lengths: Optional[dict[str, int]] = None
 
-        This method uses the gffutils package to create and in-memory database
-        of the GFF3 file.
+    def set_db(self, path: Path) -> None:
+        """Load GFF3 file into `gffutils.FeatureDB`.
 
         Args:
-            gff_file:
-                path to the input GFF3 file. If None, input is taken from the
-                standard input (stdin).
+            gff_file: Path to a GFF3 file.
         """
-        if gff_file is None:
+        try:
             self.db = gffutils.create_db(
-                sys.stdin, dbfn=":memory:", force=True, keep_order=True
+                str(path), dbfn=":memory:", force=True, keep_order=True
             )
-        else:
-            self.db = gffutils.create_db(
-                str(gff_file), dbfn=":memory:", force=True, keep_order=True
+        except gffutils.exceptions.EmptyInputError:
+            pass
+
+    def set_seq_lengths(self, path: Optional[Path] = None) -> None:
+        """Set the reference sequence lengths.
+
+        Args:
+            path: Path to a tabulated file containing the names of reference
+                sequences and the corresponding total length, in base pairs.
+        """
+        self.seq_lengths = {}
+        anno_lengths = {}
+
+        if not isinstance(self.db, gffutils.FeatureDB):
+            return
+
+        for _id in self.db.seqids():  # type: ignore
+            anno_lengths[_id] = max(
+                rec.end for rec in self.db.region(_id)  # type: ignore
             )
 
-    def extend_mirnas(
-        self,
-        primir_out: Path,
-        mir_out: Path,
-        n: int = 6,
-        seq_lengths: Optional[dict[str, int]] = None,
+        if path is None:
+            self.seq_lengths = anno_lengths
+        else:
+            with open(path, encoding="utf-8") as _file:
+                for line in _file:
+                    ref_seq, length = line.strip().split("\t")
+
+                    try:
+                        max_len = anno_lengths[ref_seq]
+                    except KeyError:
+                        max_len = 0
+
+                    try:
+                        self.seq_lengths[ref_seq] = int(length)
+                        if max_len > int(length):
+                            raise Exception(
```

Comment:
Why not raise a more specific exception. You could, e.g., create a custom exception `AnnotationException` that inherits from `ValueError`.

However, note that this statement should be outside of the `try`-`except` block - which is meant to only catch integer conversion errors. In any case it does not make sense to be inside `try`-`catch`, as the `if` clause acts as a replacement for `try`-`catch`.